site/docs/skylark/aspects.md

```diff
@@ -116,85 +150,248 @@ functions. They return [providers](rules.md#providers), can generate
+Aspects with parameters cannot be used via the command line because there is no
+syntax to define the parameters.
+
+Aspects are also allowed to have private attributes of types ``label`` or 
+``label_list``. Private label attributes can be used to specify dependencies on 
+tools or libraries that are needed for actions generated by aspects. There is not
+a private attribute defined in this example, but the following code snippet
+demonstrates how you could pass in a tool to a aspect:
+
+```python
+...
+    attrs = {
+      "_protoc" : attr.label(
+          default = Label("//tools:protoc"),
+          executable = True
+      )
+    }
+...
+```
+
+### Aspect implementation
+
+```python
+FileCount = provider(
+    fields = {
+        'count' : "number of files"
+    }
+)
+
+def _file_count_aspect_impl(target, ctx):
+    count = 0
+    # Make sure the rule has a srcs attribute.
+    if hasattr(ctx.rule.attr, "srcs"):
+        # Iterate through the sources counting files
+        sources = [f for src in ctx.rule.attr.srcs for f in src.files]
+        if ctx.attr.extension == "*":
+          count = len(sources)
+        else:
+          for source in sources:
+            if source.path.endswith(ctx.attr.extension):
+              count = count + 1
+    # Get the counts from our dependencies.
+    for dep in ctx.rule.attr.deps:
+        count = count + dep[FileCount].count
+    return [FileCount(count = count)]
+```
 
 Just like a rule implementation function, an aspect implementation function
 returns a struct of providers that are accessible to its dependencies.
 
-*   The set of providers for an aspect application A(X) is the union of providers
-    that come from the implementation of a rule for target X and from
-    the implementation of aspect A. It is an error if a target and an aspect that
-    is applied to it each provide a provider with the same name.
-*   For the aspect implementation, the values of attributes along which
-    the aspect is propagated (from the `attr_aspect` list) are replaced with
-    the results of an application of the aspect to them. For example, if target
-    X has Y and Z in its deps, `ctx.rule.attr.deps` for A(X) will be [A(Y), A(Z)].
-    In the `_metal_proto_aspect_impl` function above, ctx.rule.attr.deps will be
-    Target objects that are the results of applying the aspect to the 'deps'
-    of the original target to which the aspect has been applied.
-    That allows the aspect to examine `metal_proto` provider on them.
+In this example the ``FileCount`` is defined as a provider that has one field
+``count``. It is best practice to explicitly define the fields of a provider
+using the ``fields`` attribute.
 
+The set of providers for an aspect application A(X) is the union of providers
+that come from the implementation of a rule for target X and from
+the implementation of aspect A. It is an error if a target and an aspect that
+is applied to it each provide a provider with the same name.
```

Comment:
I'd also add that the providers that the rule implementation propagates are created and frozen before the aspect is applied and cannot be modified from the aspect.

Reply:
Fixed